Python documentation (#579)
restatedev/documentation: docs/develop/python/service-communication.mdx

    It is guaranteed that call A will execute before call B.
    It is not guaranteed though that call B will be executed immediately after call A, as invocations coming from other handlers/sources, could interleave these two calls.
</Admonition>

## Using an idempotency key

> While service-to-service communication provides *exactly once semantics*, it won't guarantee that two separate services, or a service and an external client, won't submit the same logical request twice.

Reviewer: ```suggestion
While service-to-service communication provides *exactly-once semantics*, it won't guarantee that two separate services, or a service and an external client, won't submit the same logical request twice.
```
Also change this for the other SDKs

To get this additional guarantee, you can provide an [idempotency key](https://docs.restate.dev/develop/java/clients/#invoke-a-handler-idempotently) on any request by using the parameter `idempotency_key`:

```python
CODE_LOAD::python/src/develop/service_communication.py#idempotency_key
```

## Re-attach to an invocation

> You can re-attach to any request by using the `ctx.attach_invocation` API, providing the invocation ID:

Reviewer: Do not put code in the text itself cause we will not catch it when it gets outdated. Just highlight the relevant line in the snippet below with adding `# !mark` as a line before it.

```suggestion
You can re-attach to any request via:
```

```python
CODE_LOAD::python/src/develop/service_communication.py#attach
```

<Admonition type="tip" title="Best used with idempotency key">
    Attaching to requests with an idempotency key, lets you wait for the request to finish and retrieve its response. Restate persists the response to these requests for the configured retention time.

    For requests without an idempotency key, you will not be able to retrieve the result, only wait for them to complete.
</Admonition>

## Cancel an invocation

> You can [cancel an invocation](https://docs.restate.dev/operate/invocation/#cancelling-invocations) by using the `ctx.cancel_invocation` API, providing the invocation ID:

Reviewer: Do not put code in the text itself cause we will not catch it when it gets outdated. Just highlight the relevant line in the snippet below with adding `# !mark` as a line before it. 
 ```suggestion
You can [cancel an invocation](https://docs.restate.dev/operate/invocation/#cancelling-invocations) as follows:
```
